python_modules/dagster/dagster_tests/core_tests/resource_tests/pythonic_resources/test_lifecycle_methods.py

```diff
@@ -0,0 +1,669 @@
+            log.append("setup_for_execution")
+            resources_initialized[0] += 1
+            if resources_initialized[0] == 2:
+                log.append("raising error")
+                raise Exception("my setup function errored!")
+
+        def teardown_after_execution(self, context: InitResourceContext) -> None:
+            log.append("teardown_after_execution")
+
+    class MySecondResource(ConfigurableResource):
+        def setup_for_execution(self, context: InitResourceContext) -> None:
+            log.append("setup_for_execution_second")
+            resources_initialized[0] += 1
+            if resources_initialized[0] == 2:
+                log.append("raising error")
+                raise Exception("my setup function errored!")
+
+        def teardown_after_execution(self, context: InitResourceContext) -> None:
+            log.append("teardown_after_execution_second")
+
+    @op
+    def my_never_run_op(first: MyResource, second: MySecondResource):
+        log.append("my_never_run_op")
+
+    @job(resource_defs={"first": MyResource(), "second": MySecondResource()})
+    def hello_world_job() -> None:
+        my_never_run_op()
+
+    with pytest.raises(
+        DagsterResourceFunctionError,
+        match="Error executing resource_fn on ResourceDefinition second",
+    ):
+        hello_world_job.execute_in_process()
+
+    # order is not deterministic
+    assert log == [
+        "setup_for_execution",
+        "setup_for_execution_second",
+        "raising error",
+        "teardown_after_execution",
+    ] or log == [
+        "setup_for_execution_second",
+        "setup_for_execution",
+        "raising error",
+        "teardown_after_execution_second",
+    ]
+
+
+def test_multiple_yield_ordering() -> None:
+    # ensure proper ordering of yield_for_execution
+
+    log = []
+
+    class MyResource(ConfigurableResource):
+        @contextlib.contextmanager
+        def yield_for_execution(
+            self, context: InitResourceContext
+        ) -> Generator["MyResource", None, None]:
+            log.append("yield_start_my_resource")
+            yield self
+            log.append("yield_end_my_resource")
+
+    class MySecondResource(ConfigurableResource):
+        @contextlib.contextmanager
+        def yield_for_execution(
+            self, context: InitResourceContext
+        ) -> Generator["MySecondResource", None, None]:
+            log.append("yield_start_second_resource")
+            yield self
+            log.append("yield_end_second_resource")
+
+    @op
+    def my_hello_world_op(first: MyResource, second: MySecondResource):
+        log.append("my_hello_world_op")
+
+    @job
+    def hello_world_job() -> None:
+        my_hello_world_op()
+
+    defs = Definitions(
+        jobs=[hello_world_job], resources={"first": MyResource(), "second": MySecondResource()}
+    )
+
+    assert defs.get_job_def("hello_world_job").execute_in_process().success
+
+    # order is not deterministic
```

Comment:
oh how interesting. I suppose we iterate through an unordered dictionary somewhere which causes this?